src/Tools/dotnet-openapi/src/ProjectExtensions.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+
+namespace Microsoft.DotNet.OpenApi
+{
+    public static class ProjectExtensions
+    {
+        public static ProjectItemGroupElement FindUniformOrCreateItemGroupWithCondition(this ProjectRootElement root, string projectItemElementType, string framework)
+        {
+            var lastMatchingItemGroup = FindExistingUniformItemGroupWithCondition(root, projectItemElementType, framework);
+
+            if (lastMatchingItemGroup != null)
+            {
+                return lastMatchingItemGroup;
+            }
+
+            ProjectItemGroupElement ret = root.CreateItemGroupElement();
+            if (TryGetFrameworkConditionString(framework, out string condStr))
+            {
+                ret.Condition = condStr;
+            }
+
+            root.InsertAfterChild(ret, root.LastItemGroup());
+            return ret;
+        }
+
+        public static void AddElementWithAttributes(this Project project, string tagName, string include, IDictionary<string, string> metadata)
+        {
+            var root = ProjectRootElement.Open(project.FullPath);
```

Comment:
I found a better API: call Project.AddItem(tag, file).Single() which gives a ProjectItem. You can then call ProjectItem.Xml.AddMetadata(attributename, attributevalue, expressAsAttribute).

Not sure why AddItem gives an ienumerable of items though. Maybe it's evaluating globs?

Reply:
Nice! I knew about the `AddItem` function but couldn't figure out how to make it let me add metadata as attributes.